diff -u -r imapsync-1.182/imapsync imapsync-1.182-base64-fix/imapsync
--- imapsync-1.182/imapsync	2006-08-01 16:07:42.000000000 -0700
+++ imapsync-1.182-base64-fix/imapsync	2006-09-27 16:09:59.000000000 -0700
@@ -635,7 +635,7 @@
 
         my $string = sprintf("%s\x00%s\x00%s", $imap->User,
                             $imap->Authuser, $imap->Password);
-        return encode_base64("$string");
+        return encode_base64("$string", "");
 }
 
 
